README.md
# (MODX)EvolutionCMS.snippets.ddGetDate

Snippet returns the date in a specified format. It can also:
* Return the current date;
* Return document pubdate by default, or createon if pubdate is empty;
* Display month as a string (like `January`, `February`, `March`, etc);
* Return short formated date (like `Today`, `Yesterday`, `Day before yesterday`).


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS](https://github.com/evolution-cms/evolution) >= 1.1
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.biz/modx/ddtools) >= 0.48


## Documentation


### Installation


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddGetDate`.
2. Description: `<b>2.2</b> Snippet returns the date in a specified format.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddGetDate_snippet.php` file from the archive.


#### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddGetDate/`.
2. Extract the archive to the folder (except `ddGetDate_snippet.php`).


### Parameters description

* `date`
	* Desctription: Date.
	* Valid values:
		* `integer`
		* `string`
		* `'now'` — current date
	* Default value: Pubdate or createdon (if pubdate is empty).
	
* `format`
	* Desctription: Date format to display.
	* Valid values: `string`
	* Default value: `'d.m.y'`
	
* `monthToStr`
	* Desctription: Display month as string. If it's true month in `format` must be specified as `'month'`.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `shortFormat`
	* Desctription: Display shorted fromated date. If it's true date in `shortFormat` must be specified as `'short'`.
	* Valid values: `string`
	* Default value: —
	
* `lang`
	* Desctription: Month names language.
	* Valid values:
		* `'ru'`
		* `'en'`
	* Default value: `'ru'`


### Examples


#### Get current year

```
[[ddGetDate?
	&date=`now`
	&format=`Y`
]]
```

Returns:

```
2021
```


#### Get document `pubdate` or `creadeon` (if `pubdate` is empty)

```
[[ddGetDate]]
```

Returns:

```
27.03.21
```


#### Get some TV-date with month name to string

```
[[ddGetDate?
	&date=`[*tvDate*]`
	&format=`d month Y`
	&monthToStr=`1`
	&lang=`en`
]]
```

Returns:

```
27 March 2021
```


#### Get date in short format (for example, today is 28.03.2021)

```
[[ddGetDate?
	&shortFormat=`short, G:i`
]]
```

Returns:

```
Yesterday, 9:48.
```


#### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
\DDTools\Snippet::runSnippet([
	'name' => 'ddGetDate',
	'params' => [
		'date' => 'now',
		'format' => 'd.m.Y'
	]
]);
```


## Links

* [Home page](https://code.divandesign.biz/modx/ddgetdate)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddgetdate)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddGetDate changelog


## Version 2.2 (2021-04-07)
* \* Attention! PHP >= 5.6 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.48 is required.
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ `\ddGetDate\Snippet`: The new class. All snippet code was moved here.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \+ Composer.json.


## Version 2.1.2 (2014-07-13)
* \* Bugfix: undeclared variables in PHP 5.3.
* \* The redundant code has been removed.


## Version 2.1.1 (2012-03-16)
* \+ Added English support for short formated date (if used `shortFormat`).


## Version 2.1 (2012-03-05)
* \+ Added English support in month names (if `monthToStr` equals `1`).


## Version 2.0 (2012-01-23)
* \* Attention! Backward compatibility is broken.
* \* Parameters → `date`:
	* \+ Now supports not only unixtime. Added unixtime check and converting with `strtotime()`.
	* \* Value `'current'` was replaced to `'now'` (`strtotime()` native support).


## Version 1.9 (2010-11-23)
* \* Minor optimization.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

ddGetDate.php
<?php
/** 
 * ddGetDate.php
 * @version 2.1.1 (2012-03-16)
 * 
 * @desc Snippet returns the date in a specified format.
 * 
 * @param $date {integer | string} - Date. Value 'now' — current date. Default: Pubdate or createdon (if pubdate is empty).
 * @param $format {string} - Date format to display. Default: 'd.m.y'.
 * @param $monthToStr {0; 1} - Display month as string. If it's true month in «format» must be specified as 'month'. Default: 0.
 * @param $shortFormat {string} - Display shorted fromated date. If it's true date in «shortFormat» must be specified as 'short'. Default: ''.
 * @param $lang {ru; en} - Month names language. Default: 'ru'.
 * 
 * @copyright 2012, DivanDesign
 * http://www.DivanDesign.biz
 */

$format = isset($format) ? $format : 'd.m.y';

if (!isset($date)){
	$date = ($modx->documentObject['pub_date']) ? $modx->documentObject['pub_date'] : $modx->documentObject['createdon'];
}

//Если дата не является Unix-меткой
if (!is_numeric($date)){
	$date = strtotime($date);
}

if ($date){
	if (isset($lang) && $lang == 'en'){
		$short = array('Day before yesterday', 'Yesterday', 'Today');
		$monthes = array('January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December');
	}else{
		$short = array('Позавчера', 'Вчера', 'Сегодня');
		$monthes = array('января', 'февраля', 'марта', 'апреля', 'мая', 'июня', 'июля', 'августа', 'сентября', 'октября', 'ноября', 'декабря');
	}
	
	//Если задан короткий формат и совпадает год с месяцем, то пытаемся его вывести
	if (isset($shortFormat)){
		//Если разница времени меньше чем в один день, то добавляем "Сегодня"
		if (time() - date($date) < 86400){
			return date(str_replace('short', $short[2], $shortFormat), $date);
		//Вчера
		}else if (time() - date($date) < 172800){
			return date(str_replace('short', $short[1], $shortFormat), $date);
		//Позавчера
		}else if (time() - date($date) < 259200){
			return date(str_replace('short', $short[0], $shortFormat), $date);
		}
	}
	
	if (isset($monthToStr) && $monthToStr == '1'){
		$format = str_replace('month', '\m\o\n\t\h', $format);
		$result = str_replace('month', $monthes[date('n',$date)-1], date($format, $date));
	}else{
		$result = date($format, $date);
	}
	
	return $result;
}
?>

src/Snippet.php
<?php
namespace ddGetDate;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '2.2.0',
		
		$params = [
			//Defaults
			'date' => null,
			'format' => 'd.m.y',
			'monthToStr' => false,
			'shortFormat' => null,
			'lang' => 'ru'
		]
	;
	
	/**
	 * prepareParams
	 * @version 1.0 (2021-03-27)
	 * 
	 * @param $this->params {stdClass|arrayAssociative|stringJsonObject|stringQueryFormatted}
	 * 
	 * @return {void}
	 */
	protected function prepareParams($params = []){
		//Call base method
		parent::prepareParams($params);
		
		if (is_null($this->params->date)){
			$this->params->date =
				\ddTools::$modx->documentObject['pub_date'] ?
				\ddTools::$modx->documentObject['pub_date'] :
				\ddTools::$modx->documentObject['createdon']
			;
		}
		
		//Если дата не является Unix-меткой
		if (!is_numeric($this->params->date)){
			$this->params->date = strtotime($this->params->date);
		}
	}
	
	/**
	 * run
	 * @version 1.0.1 (2021-03-27)
	 * 
	 * @return {string}
	 */
	public function run(){
		//The snippet must return an empty string even if result is absent
		$result = '';
		
		if ($this->params->date){
			if ($this->params->lang == 'en'){
				$short = [
					'Day before yesterday',
					'Yesterday',
					'Today'
				];
				
				$monthes = [
					'January',
					'February',
					'March',
					'April',
					'May',
					'June',
					'July',
					'August',
					'September',
					'October',
					'November',
					'December'
				];
			}else{
				$short = [
					'Позавчера',
					'Вчера',
					'Сегодня'
				];
				
				$monthes = [
					'января',
					'февраля',
					'марта',
					'апреля',
					'мая',
					'июня',
					'июля',
					'августа',
					'сентября',
					'октября',
					'ноября',
					'декабря'
				];
			}
			
			//Если задан короткий формат и совпадает год с месяцем, то пытаемся его вывести
			if (!is_null($this->params->shortFormat)){
				//Если разница времени меньше чем в один день, то добавляем "Сегодня"
				if (
					(
						time() -
						date($this->params->date)
					) <
					86400
				){
					$result = date(
						str_replace(
							'short',
							$short[2],
							$this->params->shortFormat
						),
						$this->params->date
					);
				//Вчера
				}elseif (
					(
						time() -
						date($this->params->date)
					) <
					172800
				){
					$result = date(
						str_replace(
							'short',
							$short[1],
							$this->params->shortFormat
						),
						$this->params->date
					);
				//Позавчера
				}elseif (
					(
						time() -
						date($this->params->date)
					) <
					259200
				){
					$result = date(
						str_replace(
							'short',
							$short[0],
							$this->params->shortFormat
						),
						$this->params->date
					);
				}else{
					//Флаг, что короткий формат не использовался
					$this->params->shortFormat = null;
				}
			}
			
			//If short format did not used
			if (is_null($this->params->shortFormat)){
				if ($this->params->monthToStr){
					$this->params->format = str_replace(
						'month',
						'\m\o\n\t\h',
						$this->params->format
					);
					
					$result = str_replace(
						'month',
						$monthes[
							date(
								'n',
								$this->params->date
							) -
							1
						],
						date(
							$this->params->format,
							$this->params->date
						)
					);
				}else{
					$result = date(
						$this->params->format,
						$this->params->date
					);
				}
			}
		}
		
		return $result;
	}
}